### ISSUE TITLE
Output tensor with stride 0 dimension gives wrong result. 

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#23063]
## 🐛 Bug

When we specify output tensor in certain torch operations (including in-place operations), in cases where the actual storage associated with the output tensor cannot hold all the output elements (i.e. the output tensor `has_internal_overlap`), the result is not correct.

The issue was found by @FDecaYed and is best described in the repro code below.

## To Reproduce
case 1. output tensor with internal_overlap storage
```
In [1]: import torch

In [2]: a = torch.empty(1) 
   ...: b = a.expand(10) 
   ...: torch.arange(10, out=b)
Out[2]: tensor([9., 9., 9., 9., 9., 9., 9., 9., 9., 9.])

In [3]: torch.arange(10)
Out[3]: tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```
case 2. in-place operation on tensor with internal_overlap storage
```
In [1]: import torch
In [2]: a = torch.empty(1)
   ...: c = torch.randn(10) 
   ...: b = a.expand(10) 
   ...: b.copy_(c)
Out[2]: 
tensor([1.6707, 1.6707, 1.6707, 1.6707, 1.6707, 1.6707, 1.6707, 1.6707, 1.6707,
        1.6707])

In [3]: c
Out[3]: 
tensor([ 1.0246, -1.0867,  0.2758,  0.2177,  0.0598,  0.8053, -0.6635, -0.6829,
         0.2282,  1.6707])
```

## Expected behavior

Well, specifying output tensor shouldn't change the output value. We need to either 1) re-allocate tensor to properly store output value; or 2) error out.

## Additional context

Two things need to be done here:
1. In the above code example, arange kernel uses `numel()` to check if the output storage is sufficient to hold output data. `numel()` does not return the actual storage space. We need to check for overlap as well.

2. The `has_internal_overlap` function could have do much better: https://github.com/pytorch/pytorch/blob/fcfefc34394a6edd8dca8a33a76307bf3f83c245/aten/src/ATen/MemoryOverlap.cpp#L10-L24.
As @ngimel kindly pointed out, we already have utility function doing the right thing: https://github.com/pytorch/pytorch/blob/86fbf17ba64f985a7d632576091a08f8344f7a24/aten/src/ATen/cuda/detail/IndexUtils.cu#L37-L71.

I'll prepare a PR shortly.

**Comment by zou3519** [comment]
@jjsjann123 you should use this helper function, if you haven't found it already, when implementing the fix: https://github.com/pytorch/pytorch/blob/84c2c89e2c08f6b0105bed1fc15d41e72bd6ea96/aten/src/ATen/MemoryOverlap.h#L19

**Comment by VitalyFedyunin** [comment]
Closed by #23574
